### PR TITLE
dynamic forward proxy: implement per filter config

### PR DESCRIPTION
**Pull request by rgs1**
This enables us to support host rewrites.

Fixes: #8607

Signed-off-by: Raul Gutierrez Segales <rgs@pinterest.com>

### CODE REVIEW

**Comment by rgs1** [comment]
I'll update the filter's doc in a bit, with an example on how to use this.

**Comment by rgs1** [comment]
cc: @fishcakez

File: api/envoy/config/filter/http/dynamic_forward_proxy/v2alpha/dynamic_forward_proxy.proto [hunk 1]
```diff
@@ -21,3 +21,10 @@ message FilterConfig {
   common.dynamic_forward_proxy.v2alpha.DnsCacheConfig dns_cache_config = 1
       [(validate.rules).message = {required: true}];
 }
+
+// Per route Configuration for the dynamic forward proxy HTTP filter.
+message PerRouteConfig {
+  // Indicates that before DNS lookup, the host header will be swapped with
```
Comment (mattklein123) [inline api/envoy/config/filter/http/dynamic_forward_proxy/v2alpha/dynamic_forward_proxy.proto]: Can you clarify what happens when this is empty? No rewrite?

File: docs/root/intro/version_history.rst [hunk 2]
```diff
@@ -39,6 +39,7 @@ Version history
 * http: :ref:`AUTO <envoy_api_enum_value_config.filter.network.http_connection_manager.v2.HttpConnectionManager.CodecType.AUTO>` codec protocol inference now requires the H2 magic bytes to be the first bytes transmitted by a downstream client.
 * http: remove h2c upgrade headers for HTTP/1 as h2c upgrades are currently not supported.
 * http: absolute URL support is now on by default. The prior behavior can be reinstated by setting :ref:`allow_absolute_url <envoy_api_field_core.Http1ProtocolOptions.allow_absolute_url>` to false.
+* http: support host rewrite in the dynamic forward proxy.
```
Comment (mattklein123) [inline docs/root/intro/version_history.rst]: ref link to the per-filter config?

Can you also add an example of the per-filter config to the main filter docs?

File: source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc [hunk 3]
```diff
@@ -62,6 +67,20 @@ Http::FilterHeadersStatus ProxyFilter::decodeHeaders(Http::HeaderMap& headers, b
     default_port = 443;
   }
 
+  // Check for per route filter config.
+  const std::string& name = HttpFilterNames::get().DynamicForwardProxy;
+  const ProxyPerRouteConfig* route_config =
+      route_entry->perFilterConfigTyped<ProxyPerRouteConfig>(name);
+  const ProxyPerRouteConfig* config =
+      route_config ? route_config
+                   : route_entry->virtualHost().perFilterConfigTyped<ProxyPerRouteConfig>(name);
```
Comment (mattklein123) [inline source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc]: I feel like there is some utility function for this pattern. Can you take a look?

File: source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.h [hunk 4]
```diff
@@ -29,6 +29,17 @@ class ProxyFilterConfig {
...
 
+class ProxyPerRouteConfig : public ::Envoy::Router::RouteSpecificFilterConfig {
+public:
+  ProxyPerRouteConfig(
+      const envoy::config::filter::http::dynamic_forward_proxy::v2alpha::PerRouteConfig& config);
+
+  const std::string& hostRewrite() const { return host_rewrite_; }
+
+private:
+  std::string host_rewrite_;
```
Comment (mattklein123) [inline source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.h]: nit: const

File: source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc (hunk 3 above)
Comment (rgs1) [reply source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc]: couldn't find one, so i added one

File: api/envoy/config/filter/http/dynamic_forward_proxy/v2alpha/dynamic_forward_proxy.proto [hunk 5]
```diff
@@ -21,3 +21,11 @@ message FilterConfig {
...
+  // this value. If this not set or empty, the original host header value
```
Comment (mattklein123) [inline api/envoy/config/filter/http/dynamic_forward_proxy/v2alpha/dynamic_forward_proxy.proto]:
```suggestion
  // this value. If not set or empty, the original host header value
```

File: docs/root/configuration/http/http_filters/dynamic_forward_proxy_filter.rst [hunk 6]
```diff
@@ -19,11 +19,10 @@ as well as the :ref:`dynamic forward proxy cluster
 must be configured together and point to the same DNS cache parameters for Envoy to operate as an
 HTTP dynamic forward proxy.
 
-.. note::
-
-  The HTTP connection manager :ref:`allow_absolute_url
-  <envoy_api_field_core.Http1ProtocolOptions.allow_absolute_url>` parameter has been set to true
-  to allow Envoy to proxy absolute HTTP URLs.
+This filter supports :ref:`host rewrite <envoy_api_msg_config.filter.http.dynamic_forward_proxy.v2alpha.FilterConfig>`
+via the per_filter_config field to provide virtual host-specific or route-specific rewrites. This
```
Comment (mattklein123) [inline docs/root/configuration/http/http_filters/dynamic_forward_proxy_filter.rst]: nit: ref link to per_filter_config

Comment (rgs1) [reply docs/root/configuration/http/http_filters/dynamic_forward_proxy_filter.rst]: done.

**Comment by mattklein123** [comment]
You are going to have to force push to fix DCO (sigh GH and suggested changes...)

/wait

**Comment by rgs1** [comment]
> You are going to have to force push to fix DCO (sigh GH and suggested changes...)
> 
> /wait

was about to ask...

**Comment by rgs1** [comment]
```
[  FAILED  ] 1 test, listed below:
[  FAILED  ] SubsetCompatibleLoadBalancers/HttpSubsetLbIntegrationTest.SubsetLoadBalancer/RANDOM, where GetParam() = 3
```

Looks unrelated...

**Pull request merged**